sstable mx reader: implement reverse single-partition reads

kbr- opened the pull request:
Until now reversed queries were implemented inside
`querier::consume_page` (more precisely, inside the free function
`consume_page` used by `querier::consume_page`) by wrapping the
passed-in reader into `make_reversing_reader` and then consuming
fragments from the resulting reversed reader.

The first couple of commits change that by pushing the reversing down below
the `make_combined_reader` call in `table::query`. This allows
working on improving reversing for memtables independently from
reversing for sstables.

We then extend the `index_reader` with functions that allow
reading the promoted index in reverse.

We introduce `partition_reversing_data_source`, which wraps an sstable data
file and returns data buffers with contents of a single chosen partition
as if the rows were stored in reverse order.

We use the reversing source and the extended index reader in
`mx_sstable_mutation_reader` to implement efficient (at least in theory)
reversed single-partition reads.

The patchset disables cache for reversed reads. Fast-forwarding
is not supported in the mx reader for reversed queries at this point.

Details in commit messages. Read the commits in topological order
for best review experience.

Refs: #9134
(not saying "Fixes" because it's only for single-partition queries
without forwarding)

kbr- commented:
Actually not every test is passing after I recompiled them again (weird - I'm sure I used `ninja dev-build` earlier)... but they did when the branch was based on Botond's reverse-reads/v3. Will debug tomorrow.

kbr- commented:
Rebased on Botond's latest v5 branch. Unit tests are passing both in dev and debug.

avikivity commented on table.cc:
```diff
@@ -2053,7 +2052,8 @@ table::mutation_query(schema_ptr s,
   try {
     auto rrb = reconcilable_result_builder(*s, cmd.slice, std::move(accounter));
-    auto r = co_await q.consume_page(std::move(rrb), cmd.get_row_limit(), cmd.partition_limit, cmd.timestamp, class_config.max_memory_for_unlimited_query);
+    assert(cmd.max_result_size); // TODO: how do we know it's engaged?
```
Was added to IDL in 92a7b16cba437501a02394e19c58f34f50ffa0be (4.3). So upgrading from 4.2 to 4.3 could break this.

Please split off into a separate minimal fix for backporting.

avikivity replied:
(/cc @denesb)

kbr- replied:
It's already a separate commit so should be easy to backport, but maybe I should replace the assertion with using a default value if this can break upgrades (also this probably means UB in master right now - since we use *cmd.max_result_size? unless it was added post-4.3 and we assume that we never upgrade from 4.2 to 4.4/higher)

avikivity replied:
It needs an issue and it needs to be treated separately from this long series, it's too confusing.

denesb replied:
We know because all entry paths in the replica-side storage proxy set it if it came unset from the coordinator (coordinator is an old node).

denesb replied:
See https://github.com/scylladb/scylla/blob/dfc135dbd1944b0ed4a98980780139770514588e/service/storage_proxy.cc#L5020-L5023

denesb replied:
Also https://github.com/scylladb/scylla/blob/dfc135dbd1944b0ed4a98980780139770514588e/service/storage_proxy.cc#L5051-L5053

avikivity commented on table.cc:
```diff
@@ -145,9 +145,12 @@ table::make_reader(schema_ptr s,
                            streamed_mutation::forwarding fwd,
-                           mutation_reader::forwarding fwd_mr) const {
+                           mutation_reader::forwarding fwd_mr,
+                           std::optional<query::max_result_size> max_memory_for_unlimited_query) const {
```
ps. this (and any other parameter that is the same for a reader graph)  is a good candidate for moving into the permit.

avikivity commented on table.cc:
```diff
@@ -180,7 +180,11 @@ table::make_reader(schema_ptr s,
     }
 
-    if (cache_enabled() && !slice.options.contains(query::partition_slice::option::bypass_cache)) {
+    if (cache_enabled() && !slice.options.contains(query::partition_slice::option::bypass_cache)
```
Please add explanation to changelog.

avikivity commented on table.cc:
```diff
@@ -180,7 +180,11 @@ table::make_reader(schema_ptr s,
+        // FIXME: reenable cache for reversed reads after:
+        // - support for reversed reads is implemented in the cache,
+        // - fast forwarding is implemented in reversed sstable readers.
         readers.emplace_back(_cache.make_reader(s, permit, range, slice, pc, std::move(trace_state), fwd, fwd_mr));
```
A more robust way of doing this is to have the cache notice a reversed read, and bypass itself. This avoids embedding knowledge about cache abilities in table.cc. Not suggesting to change, since we'll fix it soon anyway.

avikivity commented on table.cc:
```diff
@@ -187,12 +189,12 @@ table::make_reader(schema_ptr s,
-        readers.emplace_back(make_sstable_reader(s, permit, _sstables, range, slice, pc, std::move(trace_state), fwd, fwd_mr));
+        readers.emplace_back(maybe_reverse(
+                make_sstable_reader(maybe_reverse(s, slice), permit, _sstables, range, slice, pc, std::move(trace_state), fwd, fwd_mr),
+                slice, max_memory_for_unlimited_query));
```
And similarly, we could have made the sstable reader and memtable reader call make_reversing_reader() so they can undo  it locally when they gain the ability.

kbr- replied:
That's the idea. I do it in following commits for sstable reader. For memtable reader I imagine @enedil's patchset could do this when rebased on top of mine (I'll do the rebase).

avikivity replied:
Yes, I saw it later and forgot to acknowledge.

avikivity commented on database.hh:
```diff
@@ -588,6 +588,15 @@ private:
+    // and `max_memory_for_unlimited_query` must be engaged.
+    // Reversed slices must be provided in the 'half-reversed' format (the order of ranges
+    // being reversed, but the ranges themselves are not).
+    //
```
Note: reversed multi-partition reads can be useful, consider the query `SELECT game, user, score FROM tab ORDER BY score ASC PER PARTITION LIMIT 10` to select the 10 lowest scoring users for each game and prepare an advertisement for a coach for them (the table is naturally ordered by descending score since we usually want the top scores). So we should enable those queries too later.

kbr- replied:
They are enabled, just inefficient (they use `make_reversing_reader`)

avikivity replied:
Ok, fine for now.

kbr- replied:
Correction: they are not enabled, they never were.
```
ORDER BY is only supported when the partition key is restricted by an EQ or an IN.
```

Not saying we shouldn't perhaps enable them some time in the future, but at least we don't need to think about breaking potential current users which depend on such queries running today.

avikivity commented on sstables/index_reader.hh:
```diff
@@ -335,7 +335,7 @@ std::unique_ptr<clustered_index_cursor> promoted_index::make_cursor(shared_sstab
     }
 
-    if (sst->get_version() >= sstable_version_types::mc && use_binary_search_in_promoted_index) {
+    if (sst->get_version() >= sstable_version_types::mc) {
```
Well, the least intrusive change would be to move it to the constructor, not remove it completely. But I think we've proven binary search works and we can retire this parameter.

avikivity commented on sstables/mx/partition_reversing_data_source.cc:
```diff
@@ -0,0 +1,588 @@
+        co_yield read_16(*_processing_data);
+        // skip the partition key and the deletion_time
+        co_yield skip(*_processing_data, uint32_t{_u16} + sizeof(uint32_t) + sizeof(uint64_t));
+        co_yield read_8(*_processing_data);
```
We need better documentation of the field names in comments (the first read_16 is not documented), otherwise the reader has to reverse-engineer the code to understand it.

avikivity commented on sstables/mx/partition_reversing_data_source.cc:
```diff
@@ -0,0 +1,588 @@
+        co_yield skip(*_processing_data, uint32_t{_u16} + sizeof(uint32_t) + sizeof(uint64_t));
+        co_yield read_8(*_processing_data);
+        auto flags = unfiltered_flags_m(_u8);
+         if (flags.is_end_of_partition() || flags.is_range_tombstone() || !flags.has_extended_flags()) {
```
This if has a little too much indent.

avikivity commented on sstables/mx/partition_reversing_data_source.cc:
```diff
@@ -0,0 +1,588 @@
+                co_yield data_consumer::proceed::no;
+            }
+        }
+        co_yield read_unsigned_vint(*_processing_data);
```
Again document what we're ignoring here.

avikivity commented on interval.hh:
```diff
@@ -730,6 +730,15 @@ public:
 };
 
+template <typename T>
+nonwrapping_interval<T> reversed(nonwrapping_interval<T> i) {
```
Please add a comment that explains that the reversed interval must be used with a comparator that has the reverse sense of the one that was used to interpret the input.

avikivity commented on sstables/mx/reader.cc:
```diff
@@ -1107,9 +1115,9 @@ class data_consume_rows_context_m : public data_consumer::continuous_data_consum
+            // Ignore result (marker_body_size or row_body_size)
             co_yield read_unsigned_vint(*_processing_data);
-            // Ignore result
+            // Ignore result (prev_unfiltered_size)
```
Should probably be part of previous patch.

avikivity commented on test/boost/sstable_3_x_test.cc:
```diff
@@ -2860,6 +2884,60 @@ SEASTAR_THREAD_TEST_CASE(test_uncompressed_simple_read) {
+    assert_that(sst.make_reversing_reader(range))
+        .produces_partition_start(to_key(3))
+        .produces_row(clustering_key::from_single_value(*UNCOMPRESSED_SIMPLE_SCHEMA,
+                                                        int32_type->decompose(103)),
```
Need to test more complicated schemas. Are the partitions large enough that the promoted index is exercised?

avikivity replied:
(need to test both cases, and need to test a read starting at every row of a large partition so we find edge cases)

kbr- replied:
> Are the partitions large enough that the promoted index is exercised?

Probably not in this test, but it is in `test_sstable_reversing_conforms_to_mutation_source` (in `test/boost/sstable_conforms_to_mutation_source_test.cc`) as the test there is executed for different values of `sstable_writer_config::promoted_index_block_size`. It actually already managed to catch a bug which manifested for block size = 128 but not for block size =  1.

avikivity commented:
Reviewed the post-Botond patches, looks good.

denesb commented on database.cc:
```diff
@@ -1387,7 +1387,6 @@ database::query(schema_ptr s, const query::read_command& cmd, query::result_opti
     }
     column_family& cf = find_column_family(cmd.cf_id);
     auto& semaphore = get_reader_concurrency_semaphore();
-    auto class_config = query::query_class_config{.semaphore = semaphore, .max_memory_for_unlimited_query = *cmd.max_result_size};
```
Indeed this struct has no reason to exist anymore:
* `max_memory_for_unlimited_query` is always equal to `cmd.max_result_size` since this comes from the coordinator.
* The semaphore is unused since we moved admission in front of the cache.
Please also remove the struct itself, I don't think there are any more users.

kbr- commented:
v2:
- rebase on master
- use max_result_size from reader_permit in `make_reversing_reader` (from #9341). No need for the additional parameter plumbing introduced in v1
- better documentation of what we're skipping in partition_reversing_data_source + fixed indentation
- explain why we disable the cache for reversed reads in the commit message
- better assertions in mx reader constructor (use `on_internal_error` instead of `assert`)
- in reverse sstable mutation source test, don't use `make_reversing_reader`, but create the sstable reversed in the first place
- split sstable_reversing_conforms_to_mutation_source into multiple tests (one for each sstable version and block size)
- reuse Botond's utilities for reversing slices
- store non-reversed slice in memtable scanning reader so the underlying sstable reader (if used) performs a non-reversed query even after the API change (see corresponding commit - 12437352f1b4fc8ca86cb17323a08b0df1eccb37 - for details)

TODO:
- random schema + random mutations unit test
- regression test for a certain edge case (there was a small bug in the RFC which I fixed but didn't write a test for yet)

kbr- commented:
Well it seems I finally found some bugs, using random schema + random mutations unit test.

kbr- commented:
v3:
- add a random schema + random mutations unit test. I also wanted to add a separate test that catches a certain bug I found earlier, but this random test also catches it (checked by reintroducing the bug).
- fixed a bug (found thanks to above test) in `partition_reversing_data_source` where an offset obtained from the index was interpreted as absolute (from the beginning of the sstable) when it should be interpreted as relative (from the beginning of the partition). This caused weird things to happen when we queried a partition in the middle of an sstable which had a non-empty promoted index.
- removed double reversing in multishard_mutation_query.cc - this was already done on an earlier version of this branch but somehow got lost along the way.

Diff: https://github.com/scylladb/scylla/compare/6a434059e83fb85e072b4c38bbf0093f28f2bf63..9971b951d94515db0a48a1230d6ac117794893f4


The random test also found another bug, unrelated to reversing: #9352.
Wonder why the random_schema utilities (for generating random schemas and mutations) are used in so few places in our tests, they are quite powerful.

avikivity commented:
> 
> The random test also found another bug, unrelated to reversing: #9352.
> Wonder why the random_schema utilities (for generating random schemas and mutations) are used in so few places in our tests, they are quite powerful.

Order of introduction.

avikivity commented on clustering_key_filter.hh:
```diff
@@ -56,6 +56,10 @@ public:
+    // Returns all clustering ranges determined by `slice` inside partition determined by `key`.
+    // If the slice contains the `reversed` option, we assume that it is given in 'half-reversed' format
+    // (i.e. the ranges within are given in reverse order, but the ranges themselves are not reversed).
+    // The ranges will be returned in forward (increasing) order even if the slice is reversed.
```
This is great, but still ambiguous, because "forward" and "reversed" depend on the context (query order or table order). 

Please us unambiguous terms, I propose table order and query order.

kbr- replied:
> "forward" and "reversed" depend on the context (query order or table order)

Not really. The function doesn't really depend on any queries or tables, it's just a simple data transformation on a slice, it can be understood outside any context.

It's all unambiguous when we consider `schema`, which is also passed in as a parameter to this function. A slice may be 'forward', 'half-reversed', or 'native-reversed' (not used here) w.r.t a specific `schema`, and that can only be interpreted in one way.

I can try improving the comment by saying "w.r.t `schema`" explicitly.

avikivity commented on mutation_partition.cc:
```diff
@@ -610,7 +610,7 @@ mutation_partition::append_clustered_row(const schema& s, position_in_partition_
     if (!_rows.empty() && (cmp(*std::prev(i), pos) >= 0)) {
         throw std::runtime_error(format("mutation_partition::append_clustered_row(): cannot append clustering row with key {} to the partition"
-                ", last clustering row is equal or greater: {}", i->key(), std::prev(i)->key()));
+                ", last clustering row is equal or greater: {}", pos, std::prev(i)->key()));
```
Queued this.

avikivity commented on flat_mutation_reader.cc:
```diff
@@ -242,6 +242,13 @@ flat_mutation_reader make_reversing_reader(flat_mutation_reader original, query:
+    if (s.options.contains(query::partition_slice::option::reversed)) {
+        return make_reversing_reader(std::move(r), std::move(max_size));
+    }
+    return r;
```
This patch deserves special treatment too, but I'd like @denesb ack before.

tgrabiec replied:
@avikivity What special treatment are you talking about?

avikivity replied:
I previous iteration had a patch which I queued, and now this iteration is rebased over it so you can't see it. The special treatment is queuing it ahead of the series in order to unblock others (since it's not restricted to mx reader). We can keep it in the series of course.

avikivity commented on table.cc:
```diff
@@ -180,7 +180,11 @@ table::make_reader(schema_ptr s,
+            && !slice.options.contains(query::partition_slice::option::reversed)) {
+        // FIXME: reenable cache for reversed reads after:
+        // - support for reversed reads is implemented in the cache,
+        // - fast forwarding is implemented in reversed sstable readers.
```
See https://github.com/scylladb/scylla/pull/9174#discussion_r711722720

kbr- replied:
I added an option to disable the new method for sstables, falling back to the old one but with cache enabled.
By default I enabled the new method so users are encouraged to switch and all integration/system tests automatically use the new method (no need to change the config). If users really see a significant performance regression for their reversed queries, they can use the fallback.
If you think a better default would be to use the old method then inform me and I'll change it.

avikivity replied:
Given that the change can happen online, I'm okay with default on. We'll have to provide a good explanation in the release notes so people can choose to change the default on upgrade, and know what to watch for. Agree that getting the new code in use by default is good.

avikivity commented on memtable.cc:
```diff
@@ -395,6 +396,7 @@ class scanning_reader final : public flat_mutation_reader::impl, private iterato
     std::optional<dht::partition_range> _delegate_range;
     flat_mutation_reader_opt _delegate;
     const io_priority_class& _pc;
+    std::optional<query::partition_slice> _fixed_slice;
```
Good approach, until we can fix the slice when it is created.

avikivity commented on test/perf/perf_fast_forward.cc:
```diff
@@ -1961,8 +1960,6 @@ int main(int argc, char** argv) {
         }
 
-        sstables::use_binary_search_in_promoted_index = app.configuration()["use-binary-search-in-promoted-index"].as<bool>();
-
```
Queued.

avikivity commented on sstables/mx/partition_reversing_data_source.cc:
```diff
@@ -0,0 +1,608 @@
+                _current_tombstone_reversing_info->kind_offset = current_position();
+                co_yield read_8(*_processing_data);
+                _current_tombstone_reversing_info->range_tombstone_kind = bound_kind_m(_u8);
+                co_yield read_16(*_processing_data);
```
Please add comments with the names of the fields you skip, to aid the (human) reader.

kbr- replied:
But there is no skipping here. We're reading an `u16`, then assigning the result to `_ck_size`, hence we've just read the size of the clustering key.

I think I documented all the places where actual skipping (where the data is ignored) is performed.

avikivity replied:
Yes, sorry.

avikivity commented:
Series looks good, appreciate the well documented patches.

kbr- commented:
v4:
- improve commit on `clustering_key_filter::get_ranges`, mentioning table order.
- New commit at the end of the patchset, adding an option `use_new_reversed_sstable_reads`. If on, we use the new algorithm for sstables and disable the cache. If off, we perform a forward query in sstables, enable the cache, and reverse the result (so we basically use the old method).

Diff: https://github.com/scylladb/scylla/compare/9971b951d94515db0a48a1230d6ac117794893f4..53523492c80253ae8ac15ce99f82a5864e379254

denesb commented on query-request.hh:
```diff
@@ -239,11 +239,16 @@ public:
 partition_slice reverse_slice(const schema& schema, partition_slice slice);
+// Half reverse slice (forwad to legacy reverse or legacy reverse to forward).
+// Also toggles the reversed bit in `partition_slice::options`.
+partition_slice half_reverse_slice(const schema&, partition_slice);
```
These should have gone to a separate commit.

denesb commented on query-request.hh:
```diff
@@ -239,11 +239,16 @@ public:
 
 // See docs/design-notes/reverse-reads.md
+// In the following functions, `schema` may be reversed or not (both work).
+// FIXME: it would be better to have functions which don't require schema at all...
```
We require the schema because we use `slice_builder`. Removing the schema would require either over-complicating the slice builder, or reverting to allowing anybody to mutate slice internals. We don't want either.

denesb replied:
I don't like that because when we have half of the mutating methods on the slice and the other half on the slice builder. Better have all of them in one place.

kbr- replied:
Back when I implemented this commit, 100% of the methods were on the slice :P
I left this FIXME with the thought that maybe I'll convert the approach back to the slice-method-based one eventually.
But maybe I'll just be lazy and remove it.

denesb replied:
I mean *all* the mutating methods, not just the reversing related ones. And we have the slice builder for a long time now.

denesb commented on sstables/mx/reader.cc:
```diff
@@ -559,6 +559,14 @@ class mp_row_consumer_m {
             }
+            if (_slice.options.contains(query::partition_slice::option::reversed) &&
+                // we always consume whole rows (i.e. `consume_row_end` is always called) when reading in reverse,
+                // even when `consume_row_start` requested to ignore the row (TODO: explain why this happens).
```
Why not just implement skipping?

kbr- replied:
It would require translating the mx reader offsets to the data file offsets (they are different due to the "partition reversing data source" lying in between), which apparently is complicated. I'm not sure I fully understand the details too. @wmitros, the original author of the reversing data source explained:

> It should be possible to translate the data positions of the
index_reader to the positions of the intermediary data stream,
but I'm not sure we want to do that, as it add some complexity
in the index_reader and the intermediary stream. One reason
for the complexity in the input streamis that we would have two
kinds of skips: a single row skip, and a skip to a clustering
range. We know the offset of the next row, so we could check that
to differentiate them. We would also need to add an information
about the position of first clustering row and end of the last one
in the index_reader.

We do have skipping, but it's implemented differently: we advance the index and the source notices it.

BTW. I forgot about this TODO, need to update the comment in next version.

denesb commented on sstables/mx/reader.cc:
```diff
@@ -1217,32 +1227,54 @@ class mx_sstable_mutation_reader : public mp_row_consumer_reader_mx {
                             const query::partition_slice& slice,
+                            // If `slice` may die after the constructor (but before the read ends),
+                            // `store_slice` tells us to store it.
+                            bool store_slice,
```
Why not add a separate constructor which first copies the slice, then delegates to the old constructor.

kbr- replied:
> Why not add a separate constructor which first copies the slice, then delegates to the old constructor.

I'm afraid this is impossible to do in C++ :(
I assume you wanted something like this:
```
    mx_sstable_mutation_reader(shared_sstable sst,
                            schema_ptr schema,
                            reader_permit permit,
                            const dht::partition_range& pr,
                            query::partition_slice&& slice,
                            const io_priority_class& pc,
                            tracing::trace_state_ptr trace_state,
                            streamed_mutation::forwarding fwd,
                            mutation_reader::forwarding fwd_mr,
                            read_monitor& mon)
        : _stored_slice(std::move(slice))
        , mx_sstable_mutation_reader(std::move(sst), std::move(schema), std::move(permit),
                pr, *_stored_slice, pc, std::move(trace_state), fwd, fwd_mr, mon)
    {}
```

denesb replied:
Yes.

kbr- replied:
In the end I used a single constructor that takes a `value_or_reference<query::partition_slice>`. Check it out.

denesb commented on sstables/mx/reader.cc:
```diff
@@ -1369,26 +1401,59 @@ class mx_sstable_mutation_reader : public mp_row_consumer_reader_mx {
+                // we use the non-reversed schema in the index_reader, so we need to reverse it again
+                bound_weight new_weight = pos.get_bound_weight() == bound_weight::before_all_prefixed ? bound_weight::after_all_prefixed
+                                        : pos.get_bound_weight() == bound_weight::after_all_prefixed ? bound_weight::before_all_prefixed
+                                        : bound_weight::equal;
```
Don't we have some `reverse(bound_weight)` now for this?

enedil replied:
I use it in memtable reader: https://github.com/scylladb/scylla/pull/9174/commits/6eab8baf9acd1885507d487b51d4f457d2fa65ab

denesb commented on sstables/mx/reader.cc:
```diff
@@ -1369,26 +1401,59 @@ class mx_sstable_mutation_reader : public mp_row_consumer_reader_mx {
+                                        : pos.get_bound_weight() == bound_weight::after_all_prefixed ? bound_weight::before_all_prefixed
+                                        : bound_weight::equal;
+                const clustering_key_prefix* new_key = !pos.has_key() ? nullptr : &pos.key();
+                auto inv_pos = position_in_partition_view(pos.region(), new_weight, new_key);
```
Or even for `position_in_partition_view`?

enedil replied:
Same as above - https://github.com/scylladb/scylla/pull/9174/commits/6eab8baf9acd1885507d487b51d4f457d2fa65ab

denesb commented on sstables/mx/reader.cc:
```diff
@@ -1369,26 +1401,59 @@ class mx_sstable_mutation_reader : public mp_row_consumer_reader_mx {
+                        }
+                    });
+                });
+            }
```
This would have benefited from coroutinizing greatly.

denesb commented on sstables/mx/reader.cc:
```diff
@@ -1417,7 +1485,14 @@ class mx_sstable_mutation_reader : public mp_row_consumer_reader_mx {
         if (_single_partition_read) {
             _read_enabled = (begin != *end);
-            _context = data_consume_single_partition<DataConsumeRowsContext>(*_schema, _sst, _consumer, { begin, *end });
+            if (_slice.options.contains(query::partition_slice::option::reversed)) {
```
This should have been saved in a `_reversed` member, to make it easier on the eyes.

denesb commented on sstables/sstables.cc:
```diff
@@ -2138,12 +2142,27 @@ sstable::make_reader_v1(
+            // Currently this can happen in only one scenario: we perform a reversed query,
+            // thus provided the reader a reversed slice and a reversed schema, but the reader rejected
+            // the reversing request and tells us to perform the reversing ourselves (the data from the reader
+            // will return in forward order).
```
Under what conditions can the reader (silently) *reject* to do what it is asked to do?

kbr- replied:
It's not silent, it's explicitly specified in the interface of `mx::make_reader`:
```
// Precondition: if the slice is reversed, the schema must be reversed as well.
// Reversed slices must be provided in the 'half-reversed' format (the order of ranges
// being reversed, but the ranges themselves are not).
//
// If the slice is reversed then:
// - if this is a single-partition read (range.is_singular()), each partition
//   in the returned fragment stream will be reversed - i.e. ordered according to
//   the reversed schema. The reader's schema will be the provided schema.
//   In this mode fast-forwarding is not supported (FIXME);
// - otherwise, the data will be returned in non-reversed order and the reader's
//   schema will be non-reversed - i.e. it will be the reverse of the provided schema
//   (since the provided schema is already reversed according to the precondition).
//   In this case the caller is responsible for reversing the fragment stream
//   themselves.
flat_mutation_reader_v2 make_reader(
```
(see sstables/mx/reader.hh) which also answers your question.

denesb replied:
I think this is suprising and we should do this here instead create a reversing reader and pass non-reversed slice if the case.

kbr- replied:
That's what I wanted to do, but found that I can't (at least I can't do it easily) because there is nowhere to store the slice, and I didn't want to stuff another intermediate reader in between (there are too many of those already for my taste). If I do this directly in the mx reader, it's easy to store the slice as you can see: in the reader itself.

I think it shouldn't cause any surprises if one reads documentation comments which I diligently wrote. And hopefully we'll be able to remove these workarounds soon, allowing us to simplify the interface.

kbr- replied:
Your suggestion in the other comment gave me an idea: provide two `make_reader` functions, one with `const partition_slice&` param, one with `partition_slice&&` which would store the slice. Both would require that if the range is not singular, the slice is not reversed (and we would have an assertion inside), and it's kinda sad that we would have two functions, but maybe it's better than the current complex interface alternative. What do you think?

denesb replied:
Yes, this is what I wanted to suggest too.

denesb commented:
Minor comments only.

kbr- commented:
v5:
- remove some leftover noise from previous versions
- typo fix
- remove unnecessary std::optional initialization
- move `half_reverse_slice` introduction to a separate commit and remove a related FIXME
- introduce a small utility, `value_or_reference<T>`, which is a `const T&` with optional `T` storage, so it can either refer to an external object or to internally stored one. Change the `mx_sstable_mutation_reader` to take `value_or_reference<query::partition_slice>` instead of `const query::partition_slice&` with `bool store_slice` so the user can ask the reader to store a slice without the constructor having to perform an additional copy.
- simplify the interface of `mx::make_reader`. Previously, for non-singular ranges and reversed slices it would internally unreverse the slice and perform a forward query. Now it is included in the precondition of `mx::make_reader` that it cannot perform non-singular reverse reads. Instead the caller (`sstable::make_reader`) unreverses the slice and passes it into `mx::make_reader`. This uses a second version of `mx::make_reader` which takes a `query::partition_slice&&` instead of `const query::partition_slice&`; this version stores the passed-in slice. 
- elaborate about why we have to call the filter for a row twice in `mp_row_consumer::consume_row_end`, removing a leftover TODO
- add private method `mx_sstable_mutation_reader::reversed() const` and use it for readability

diff: https://github.com/scylladb/scylla/compare/53523492c80253ae8ac15ce99f82a5864e379254..6f1b5ed568e674ec87d4f9ee9eaab1f2438d62a9

tgrabiec commented on querier.hh:
```diff
@@ -153,7 +134,7 @@ public:
         , _range(std::make_unique<const dht::partition_range>(std::move(range)))
         , _slice(std::make_unique<const query::partition_slice>(std::move(slice)))
-        , _reader(ms.make_reader(underlying_schema(), _permit, *_range, *_slice, pc, std::move(trace_ptr), streamed_mutation::forwarding::no, mutation_reader::forwarding::no))
+        , _reader(ms.make_reader(_schema, _permit, *_range, *_slice, pc, std::move(trace_ptr), streamed_mutation::forwarding::no, mutation_reader::forwarding::no))
```
_schema is reversed for reversed reads, right?

If we're doing this, then we have to change all mutation_sources to accept a reversed schema for a reversed read, not just table::make_reader(), so that the querier can work with any mutation_source, not only the ones which are currently passed to it.

kbr- commented:
v6:
- rebased on master, resolving merge conflicts.

tgrabiec commented on database.hh:
```diff
@@ -650,6 +650,14 @@ public:
+    //
+    // If the slice is reversed, each partition returned by the reader will be reversed,
+    // i.e. the fragments will be returned according to the order of the reversed schema,
+    // and the reader does not support fast forwarding.
```
table->as_mutation_source() calls make_reader() as is, so whatever constraints you add here, you should also add to the generic mutation_source interface, or make necessary transformations inside the adaptor in as_mutation_source().

kbr- commented:
v7:
- rebased on top of Botond's PR (https://github.com/scylladb/scylla/pull/9384).

My PR starts from commit `clustering_key_filter: document clustering_key_filter_ranges::get_ranges`. This version is simpler since the work of pushing down reversing to source level is done by Botond's changes. I also removed redundant tests (also covered by Botond's changes and partially by the random test in last commit).

The version no longer contains the configuration option for switching between reads with cache but old reversing vs reads without cache but new reversing. Now:
- if the user does `bypass cache`, the new version is executed
- otherwise, the old version is executed with cache.

kbr- commented:
Note: the dtest
`reversed_queries_test.py:TestReversedQueriesMemoryUsage.test_memory`
does not pass with this, since the new reversing code is only executed with `bypass cache` which the test does not use.

If we agree that we want the new UI, we will need to modify the test.
cc @piodul 

@avikivity what do you think of the new UI?
> The version no longer contains the configuration option for switching between reads with cache but old reversing vs reads without cache but new reversing. Now:
> - if the user does `bypass cache`, the new version is executed
> - otherwise, the old version is executed with cache.

tgrabiec commented:
I don't think we should rely on "bypass cache" to enable native reverse
reads because eventually when cache supports reversing (soon) we will want
users to benefit and use the cache with native reads. We probably want an
independent option for disabling native reverse reads.

On Tue, Sep 28, 2021 at 4:53 PM Kamil Braun ***@***.***>
wrote:

> Note: the dtest
> reversed_queries_test.py:TestReversedQueriesMemoryUsage.test_memory
> does not pass with this, since the new reversing code is only executed
> with bypass cache which the test does not use.
>
> If we agree that we want the new UI, we will need to modify the test.
> cc @piodul <https://github.com/piodul>
>
> @avikivity <https://github.com/avikivity> what do you think of the new UI?
>
> The version no longer contains the configuration option for switching
> between reads with cache but old reversing vs reads without cache but new
> reversing. Now:
>
>    - if the user does bypass cache, the new version is executed
>    - otherwise, the old version is executed with cache.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylla/pull/9281#issuecomment-929307017>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACFILZ4TTIGJPGYLA3E43DUEHJFBANCNFSM5DJHZPZQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

denesb commented:
I agree, I think relying on bypass cache for controlling this is too subtle. On the other hand @avikivity asked for the switch to enable currently fine-working small reverse reads to be able to use cache. With the solution proposed by @kbr- we achieve that: normal reads will use the old code path through cache, reads that bypass cache use the native reverse format. Once cache has native reverse support the above mentioned small reads seamlessly switch to using it.

avikivity commented:
It requires people to change their applications twice.

kbr- commented:
I readded the option from previous version, although it has a different name now: `reversed_reads_auto_bypass_cache`.

So the logic right now is:
- if the user queries with `bypass cache`, new reversing is used
- if the user queries without `bypass cache` but the `reversed_reads_auto_bypass_cache` option is on, new reversing is used and cache is bypassed
- otherwise, old reversing is used and cache is not bypassed

kbr- commented:
Rebased onto master.

kbr- commented:
Rebased again, fixed minor conflict in includes.

kbr- commented:
@denesb @tgrabiec please approve/disapprove/rereview if necessary.

tgrabiec commented:
@kbr- I'm in the middle of review.

tgrabiec commented on sstables/mx/partition_reversing_data_source.cc:
```diff
@@ -0,0 +1,608 @@
+                    co_yield read_8(*_processing_data);
+                    // `row_body_skipping_context` should not be constructed on static rows
+                    auto extended_flags = unfiltered_extended_flags_m(_u8);
+                    assert(!extended_flags.is_static());
```
on_internal_error() is more appropriate, we don't want to bring the cluster down on poisoned read

kbr- replied:
We need a version of `on_internal_error` which can be used like an `assert` (without a separate `if`):
```
assert_internal(condition, logger, error_message)
```
which is equivalent to
```
if (!condition) on_internal_error(logger, error_message)
```
it would probably have to be a macro so we don't eagerly evaluate the `error_message` expression.

tgrabiec commented on sstables/mx/partition_reversing_data_source.cc:
```diff
@@ -0,0 +1,608 @@
+                    }
+                    move_to_next_ck_block();
+                }
+                _reading_partition_ck = false;
```
"partition ck" is a bit confusing. Partitions don't have clustering keys any more than they have range tombstone keys. They have rows, and rows have clustering keys.

tgrabiec commented on sstables/mx/partition_reversing_data_source.cc:
```diff
@@ -0,0 +1,608 @@
+            } else if (flags.is_range_tombstone()) {
+                _current_tombstone_reversing_info.emplace();
+                _current_tombstone_reversing_info->kind_offset = current_position();
+                co_yield read_8(*_processing_data);
```
Seems inefficient to yield after each byte unconditionally. Performance is not important for reverse reads, so fine. But is there a reason for this?

wmitros replied:
Because of how coroutines work in c++, we were able to add a condition to the `processing_result_generator`, so that the execution of the coroutine is not suspended despite using `co_yield` keyword (`struct read_awaiter` in processing_result_generator.hh)

tgrabiec replied:
Right

kbr- commented on sstables/sstables.cc:
```diff
@@ -2125,6 +2125,10 @@ sstable::make_reader(
+
+    if (slice.options.contains(query::partition_slice::option::reversed)) {
+        on_internal_error(sstlog, "sstable::make_reader (v2 version): reversed slices not supported for sstable versions < mc");
+    }
```
I "ensured" it by writing a comment on `sstable::make_reader` declaration:
```
    // Precondition: if sstable version is < mc, the slice is not reversed.
```

I can insert a reversing wrapper. Didn't because `make_reader` is only used in tests currently (production code uses `make_reader_v1`) and the codepath is dead but I guess it's better to have it conform to the general interface.

I'm also wondering what to do with reversed multi-partition queries for >= mc versions. I can also insert a reversing wrapper, but that will require something like:
```
upgrade_to_v2(make_reversing_reader(downgrade_to_v1(mx::make_reader(...)...)
```
ugly but I can do it, it doesn't really matter anyway since reversed multi-partition queries (available from Thrift only) are effectively dead.

tgrabiec replied:
We're transitioning to v2 so soon the production code will be switched to use `make_reader()`. Better fix it now since the logic is hot in your cache.

Regarding multi-partition reversing, for now inserting converters is fine. Eventually make_reversing_reader() should be v2 natively and the converters will go away.

kbr- commented on conf/scylla.yaml:
```diff
@@ -510,3 +510,20 @@ commitlog_total_space_in_mb: -1
+# and sstable reverse read algorithms are updated to handle reversed
+# queries correctly.
+# The option can be updated without restarting Scylla.
+# reversed_reads_auto_bypass_cache: true
```
> "do not use reverse readers" (at all)

What would that do? What would happen if the user requested a reversed read from CQL? Error?

I thought the intent of this option was for the user to be able to use the new efficient reversing code seamlessly in their applications, which requires bypassing the cache at this point. So in order for them to not have to insert `bypass cache` everywhere in their queries (and then in the future remove it) I added an option to do it automatically.

tgrabiec replied:
If reverse readers are disabled by users, the request would be executed using the old logic, by querying in forward mode and reversing in memory. 

I thought the reason for this option was that this performs better than using reversed readers if the partition is small. Another reason could be to workaround bugs in the new reversed readers. 

Maybe a global option for bypassing cache for reversed readers also makes sense, even if cache supports reverse reads. I do hope that we will manage to release reversed reader support in cache together with everything. But in case of bugs it may be good to have it.

kbr- commented on sstables/mx/partition_reversing_data_source.cc:
```diff
@@ -0,0 +1,608 @@
+        _ck_blocks_header_offset = 0u;
+    }
+    bool no_more_ck_blocks() const { return _ck_column_value_fix_lengths.empty(); }
+    void move_to_next_ck_block() {
```
Not able to say after a quick glance, would have to do some deeper research. I'd suggest to leave nontrivial changes such as this for a followup considering this PR already went through 10 versions and we're already late with estimated delivery time, and it would also require modifying the code that @wmitros prepared for backporting to enterprise (a version of this code which does not use coroutines)
(also, the mx reader code which this code is based off of also does not use `clustering_parser`)

tgrabiec replied:
We can leave this for a follow up.
